File: on1y/books/translator_match.py
"""Fuzzy translator matching between Douban hints and download candidates."""

from __future__ import annotations

import re
from typing import Any

from on1y.books.display_name import parse_translator_from_text
from on1y.books.edition_match import candidate_text

_PUNCT_RE = re.compile(r"[/／,，、;；|]+")
_TRAILING_YI_RE = re.compile(r"\s*译\s*$")
_TRAILING_DENG_RE = re.compile(r"等\s*$")
_SPACE_RE = re.compile(r"\s+")
_PUBLISHER_RE = re.compile(r"出版社|出版集团|出版公司|书局|书社|Press|Publishing", re.I)


def _looks_like_publisher(text: str | None) -> bool:
    t = (text or "").strip()
    if not t:
        return False
    if _PUBLISHER_RE.search(t):
        return True
    if re.fullmatch(r"\d{4}([-/年]\d{1,2})?", t):
        return True
    return False


def _norm(text: str | None) -> str:
    return _SPACE_RE.sub("", (text or "").strip().lower())


def candidate_blob(candidate: dict[str, Any]) -> str:
    return candidate_text(
        str(candidate.get("title") or ""),
        {
            "author": candidate.get("author"),
            "publisher": candidate.get("publisher"),
            "year": candidate.get("year"),
            "language": candidate.get("language"),
        },
    )


def _clean_token(text: str) -> str:
    token = (text or "").strip()
    token = _TRAILING_YI_RE.sub("", token)
    token = _TRAILING_DENG_RE.sub("", token).strip()
    return token


def extract_translator_tokens(translator: str | None) -> list[str]:
    if not translator or not str(translator).strip():
        return []
    raw = str(translator).strip()
    parts = [p.strip() for p in _PUNCT_RE.split(raw) if p.strip()]
    if not parts:
        parts = [raw]
    tokens: list[str] = []
    for part in parts:
        cleaned = _clean_token(part)
        if len(cleaned) >= 2:
            tokens.append(cleaned)
    return tokens


def _tokens_overlap(left: str, right: str) -> bool:
    a = _norm(left)
    b = _norm(right)
    if not a or not b:
        return False
    if a in b or b in a:
        return True
    # 唐日松 vs 唐日松等
    if len(a) >= 2 and a in b:
        return True
    if len(b) >= 2 and b in a:
        return True
    return False


def translator_matches_hint(hint: str | None, candidate: dict[str, Any]) -> bool:
    """True when Douban translator hint is compatible with the picked file metadata."""
    if not hint or not str(hint).strip():
        return True

    hint_raw = str(hint).strip()
    if _looks_like_publisher(hint_raw):
        return True

    blob = candidate_blob(candidate)
    blob_norm = _norm(blob)
    hint_norm = _norm(hint_raw)

    if hint_norm in blob_norm:
        return True

    for token in extract_translator_tokens(hint_raw):
        if _norm(token) in blob_norm:
            return True

    parsed = parse_translator_from_text(
        candidate.get("title"),
        candidate.get("author"),
        candidate.get("publisher"),
    )
    if parsed:
        if _tokens_overlap(hint_raw, parsed):
            return True
        for token in extract_translator_tokens(hint_raw):
            if _tokens_overlap(token, parsed):
                return True

    return False

File: frontend/src/lib/translator-match.ts
import type { BookAcquireCandidate } from "@/lib/book-types";

const PUNCT_RE = /[/／,，、;；|]+/;
const PUBLISHER_RE = /出版社|出版集团|出版公司|书局|书社|Press|Publishing/i;

function looksLikePublisher(text: string | null | undefined): boolean {
  const t = text?.trim();
  if (!t) return false;
  return PUBLISHER_RE.test(t) || /^\d{4}([-/年]\d{1,2})?$/.test(t);
}

function norm(text: string): string {
  return text.replace(/\s+/g, "").toLowerCase();
}

function cleanToken(text: string): string {
  return text
    .trim()
    .replace(/\s*译\s*$/, "")
    .replace(/等\s*$/, "")
    .trim();
}

function extractTranslatorTokens(translator: string | null | undefined): string[] {
  const raw = translator?.trim();
  if (!raw) return [];
  const parts = raw.split(PUNCT_RE).map((p) => p.trim()).filter(Boolean);
  const chunks = parts.length > 0 ? parts : [raw];
  return chunks.map(cleanToken).filter((t) => t.length >= 2);
}

function parseTranslatorFromCandidate(candidate: BookAcquireCandidate): string | null {
  const blob = [candidate.title, candidate.author, candidate.publisher].filter(Boolean).join(" ");
  const semi = blob.match(/著\s*;\s*([^;，,]+?)\s*译/);
  if (semi?.[1]) {
    const name = semi[1].trim();
    if (name.length >= 2 && name.length <= 30) return name;
  }
  const plain = blob.match(/([\u4e00-\u9fff·A-Za-z]{2,30})\s*译/g);
  if (!plain) return null;
  for (const match of plain) {
    const m = match.match(/([\u4e00-\u9fff·A-Za-z]{2,30})\s*译/);
    const name = m?.[1]?.trim();
    if (name && name !== "翻" && name !== "编") return name;
  }
  return null;
}

function tokensOverlap(left: string, right: string): boolean {
  const a = norm(left);
  const b = norm(right);
  if (!a || !b) return false;
  return a.includes(b) || b.includes(a);
}

export function translatorMatchesHint(
  hint: string | null | undefined,
  candidate: BookAcquireCandidate
): boolean {
  const raw = hint?.trim();
  if (!raw) return true;
  if (looksLikePublisher(raw)) return true;

  const blob = [candidate.title, candidate.author, candidate.publisher].filter(Boolean).join(" ");
  const blobNorm = norm(blob);
  const hintNorm = norm(raw);

  if (hintNorm && blobNorm.includes(hintNorm)) return true;

  for (const token of extractTranslatorTokens(raw)) {
    if (blobNorm.includes(norm(token))) return true;
  }

  const parsed = parseTranslatorFromCandidate(candidate);
  if (parsed) {
    if (tokensOverlap(raw, parsed)) return true;
    for (const token of extractTranslatorTokens(raw)) {
      if (tokensOverlap(token, parsed)) return true;
    }
  }

  return false;
}

export function candidateDoubanMismatch(
  doubanTranslator: string | null | undefined,
  candidate: BookAcquireCandidate
): boolean {
  const tr = doubanTranslator?.trim();
  if (!tr) return false;
  return !translatorMatchesHint(tr, candidate);
}
